enhancements/installer/assisted-installer.md

```diff
@@ -0,0 +1,330 @@
+
+The target user is someone wanting to deploy OpenShift, especially on
+bare metal, with as little up-front infrastructure dependencies as
+possible. This person has access to servers and wants to run workloads
+quickly. They do not necessarily have the administrative privileges to
+create private VLANs, configure DHCP/PXE servers, or manage other
+aspects of the infrastructure surrounding the hardware where the
+cluster will run. They are willing to accept that the cluster they
+build may not have all of the same features as an IPI cluster on day
+1, but that by taking additional steps later they will be able to make
+it fully production-ready.
+
+### Goals
+
+- Make initial deployment of usable and supportable clusters simpler.
+- Move more infrastructure configuration from day 1 to day 2.
+- Support connected and disconnected on-premise deployments.
+- Support existing IPI features, especially for day 2 cluster
+  management and scale-out.
+
+### Non-Goals
+
+- Because the initial focus is for bare metal, this enhancement does
+  not exhaustively cover variations needed to offer similar features
+  on other platforms (such as changes to image formats, the way a host
+  boots, etc.). It is desirable to support those platforms, but that
+  work will be described separately.
+- Replace the existing OpenShift installer.
+- Describe how these workflows would work for multi-cluster
+  deployments managed with Hive or ACM.
+
+## Proposal
+
+There are several separate changes to enable the assisted installer
+workflows, including a GUI front-end for the installer, a cloud-based
+orchestration service, and changes to the installer and bootstrapping
+process.
+
+### Connected Deployments
+
+The process starts when the user goes to an "assisted installer"
+application running on `cloud.redhat.com`, enters details needed by
+the installer (pull secret, OpenShift version, ssh keys, proxy
+settings, etc.), and then downloads a live RHCOS ISO image with the
+software and settings they need to complete the installation locally.
+
+The user then boots the live ISO on each host they want to be part of
+the cluster (control plane and workers). They can do this by hand
+using thumb drives, by attaching the ISO using virtual media support
+in the BMC of the host, or any other way they choose.
+
+When the ISO boots, it starts an agent that communicates with the REST
+API for the assisted installer service running on `cloud.redhat.com`
+to receive instructions. The agent identifies itself based on the
+settings in the ISO and hardware details it discovers from the host it
+is running on. Communication always flows from agent to service (using
+polling, long polling, web sockets, or some other similar technique)
+so that firewalls and proxies work as expected.
+
+As agents report to the assisted installer, their hosts appear in the
+UI and the user is given an opportunity to examine the hardware
+details reported and to set the role and cluster of each host.
```

Comment:
I think we should try to decouple static validation and generation of defaults (which are already done in the installer, so we can add to those?), and runtime validation e.g connectivity and latency - I suspect the OpenShift team would prefer if runtime validation is added in a way that's not tied to a specific platform, e.g perhaps there should be some monitoring which exists for all platforms that does that?

Reply:
> When you say "that the VIPs are on the same machine network" which VIPs do you mean?

API and Ingress VIPs

Reply:
> > When you say "that the VIPs are on the same machine network" which VIPs do you mean?
> 
> API and Ingress VIPs

Note that this particular validation already exists in the installer https://github.com/openshift/installer/blob/master/pkg/types/baremetal/validation/platform.go#L30 so we'll need to ensure we don't duplicate existing validations in the wrapper around the installer.

Reply:
Maybe eventually, but at least for now when the installer and installer-service are totally separate, we need to be able to validate input before we actually start and fail the installation.

Reply:
> Maybe eventually, but at least for now when the installer and installer-service are totally separate, we need to be able to validate input before we actually start and fail the installation.

Note that the validation doesn't necessarily imply starting the installation, you'd just have to run the installer to generate the manifests or ignition-configs, and deal with any error.

Another option which avoids code duplication would be to vendor the installer into the install service I guess, and reuse the same validation and IP generation code.

Reply:
Since bare metal is the only platform with a concept of host inventory that exists prior to forming a cluster, I think it's natural to have some validation that is specific to bare metal and happens prior to installation or even day-2 provisioning. It's especially relevant in ACM where a user may be able to eventually choose which cluster they want to allocate a given host to, and it would be useful to validate connectivity.

Reply:
I reworked this section on validations a little, now lines 126-134.

It's not clear to me how the remote application validates the VIPs based on discovered hardware, but maybe I misunderstood that part. Maybe the VIPs are something the user gives the app?